pkg/retainer/txretainer/transaction_retainer_test.go

```diff
@@ -0,0 +1,418 @@
+	TransactionID          iotago.TransactionID
+	ValidSignature         bool
+	EarliestAttachmentSlot iotago.SlotIndex
+	State                  api.TransactionState
+	TransactionError       error
+}
+
+type test struct {
+	name          string
+	transactionID iotago.TransactionID
+	newTxMetaFunc func() *testTxMetadata
+	testFunc      func()
+	resultFunc    func() *api.TransactionMetadataResponse
+	targetErr     error
+}
+
+func (test *test) Run(t *testing.T, tr *txretainer.TransactionRetainer) {
+	var newTxMeta *testTxMetadata
+	if test.newTxMetaFunc != nil {
+		newTxMeta = test.newTxMetaFunc()
+	}
+
+	if newTxMeta != nil {
+		// update the transaction metadata in the retainer
+		tr.UpdateTransactionMetadata(newTxMeta.TransactionID, newTxMeta.ValidSignature, newTxMeta.EarliestAttachmentSlot, newTxMeta.State, newTxMeta.TransactionError)
+	}
+
+	if test.testFunc != nil {
+		test.testFunc()
+	}
+
+	// get the transaction metadata from the retainer
+	txMeta, err := tr.TransactionMetadata(test.transactionID)
+	if test.targetErr != nil {
+		require.ErrorIs(t, err, test.targetErr, test.name)
+		return
+	}
+	require.NoError(t, err, test.name)
+
+	var targetTxMeta *api.TransactionMetadataResponse
+	if test.resultFunc != nil {
+		targetTxMeta = test.resultFunc()
+	}
+	require.Equal(t, targetTxMeta, txMeta, test.name)
+}
+
+func TestTransactionRetainer_UpdateMetadata(t *testing.T) {
+	ts := newTestSuite(t)
+	defer ts.Close()
+
+	tr := ts.TxRetainer
+
+	txID := tpkg.RandTransactionID()
+
+	// IMPORTANT: all those tests modify the same transaction metadata and depend on each other
+	for _, test := range []*test{
+		// fail - transaction metadata not found because it was not added yet
+		{
+			name:          "fail - transaction metadata not found because it was not added yet",
+			transactionID: txID,
+			newTxMetaFunc: nil,
+			testFunc:      nil,
+			resultFunc: func() *api.TransactionMetadataResponse {
+				return &api.TransactionMetadataResponse{
+					TransactionID: txID,
+				}
+			},
+			targetErr: txretainer.ErrEntryNotFound,
+		},
+		// ok - add attachment with invalid signature and no attachment slot
+		{
+			name:          "ok - add attachment with invalid signature and no attachment slot",
+			transactionID: txID,
+			newTxMetaFunc: func() *testTxMetadata {
+				return &testTxMetadata{
+					TransactionID:          txID,
+					ValidSignature:         false,
+					EarliestAttachmentSlot: 0,
+					State:                  api.TransactionStateFailed,
+					TransactionError:       iotago.ErrInputAlreadySpent,
+				}
+			},
+			testFunc: nil,
+			resultFunc: func() *api.TransactionMetadataResponse {
+				return &api.TransactionMetadataResponse{
+					TransactionID:             txID,
+					TransactionState:          api.TransactionStateFailed,
+					EarliestAttachmentSlot:    0,
+					TransactionFailureReason:  api.DetermineTransactionFailureReason(iotago.ErrInputAlreadySpent),
+					TransactionFailureDetails: "input already spent",
```

Comment:
I think it's better to insert a reference rather than a copy here. If we ever update the error message in iota.go we would also have to touch iota-core with the current code.

```suggestion
					TransactionFailureDetails: iotago.ErrInputAlreadySpent.Error(),
```

Ideally do the same for the rest of the file as well.